utils/FileHandler.cpp

```diff
@@ -0,0 +1,315 @@
+        exception.errorWindow();
+    }
+
+    return shapesToReturn;
+}
+
+Shape* FileHandler::extractShape(QTextStream& qin)
+{
+    QString shapeType;  // Name of the shape being extracted currently
+    QList<int> dimensionsList;   // List of dimensions of the shape
+    Shape* shape;    // Pointer to the shape that is returned by the function
+
+    // Ignore the shape's ID data
+    qin.readLine();
+
+    // Get the name of the shape to add
+    getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    qin >> shapeType;
+
+    // Throw an exception if the shape type extracted is not in the list of valid shape types
+    if(jconstants::TYPE_LIST.indexOf(shapeType) == -1)
+    {
+        QFile* file = dynamic_cast<QFile*>(qin.device());
+
+        if(file != nullptr)
+        {
+            throw BadFileFormat(*file, "a valid shape type name", shapeType);
+        }
+        else
+        {
+            throw BadFileFormat("a valid shape type name", shapeType);
+        }
+    }
+
+    // Extract the dimensions from the file
+    dimensionsList = extractDimensions(qin, shapeType);
+
+    // If the shape being read in isn't text...
+    if(shapeType != jconstants::TEXT)
+    {
+        //...call corresponding extraction function
+        shape = extractNonText(qin, shapeType, dimensionsList);
+    }
+    // Otherwise, if the shape is a text...
+    else
+    {
+        //...call the corresponding extraction function
+        shape = extractText(qin, dimensionsList);
+    }
+
+    // Read two more lines to get to the next shape's data
+    qin.readLine();
+    qin.readLine();
+
+    return shape;
+}
+
+Shape* FileHandler::extractNonText(QTextStream& qin, QString& shapeType, QList<int>& dimensionsList)
+{
+    Shape* shape;    // Shape pointer to be returned by the function
+
+    // Extract data for the pen that draws this shape' edges
+    extractPen(qin);
+
+    // If the shape being read in isn't a line, we know it needs a brush
+    if(shapeType != jconstants::LINE && shapeType != jconstants::POLYLINE)
+    {
+        // Extract data for the brush used to fill the insides of this shape
+        extractBrush(qin);
+        shape = constructShape(shapeType, dimensionsList);
+    }
+    else
+    {
+        shape = constructLine(shapeType, dimensionsList);
+    }
+
+    return shape;
+}
+
+Shape *FileHandler::extractText(QTextStream &qin, QList<int> &dimensionsList)
+{
+    QString body;   // Body of the text
+    QFont font; // Font of the text
+
+    // Get to the next whitspace, then read in the line as the text's body
+    jbrush::getTo(qin, jconstants::TITLE_INTERRUPTOR);
+    body = qin.readLine();
+
+    // For now, ignore the shape color and shape alignment
+    qin.readLine();
+    qin.readLine();
+
+    // Extract the text's font
+    font = extractFont(qin);
+
+    return new Text(QPoint(dimensionsList[0], dimensionsList[1]), body, font);
+}
+
+QList<int> FileHandler::extractDimensions(QTextStream &qin, const QString& shapeType)
```

Comment:
References loves their types, don't let them be with the ugly variable
```suggestion
QList<int> FileHandler::extractDimensions(QTextStream& qin, const QString& shapeType)
```